diff --git a/Components/bsp/srf06eb_cc2538/drivers/source/bsp.h b/Components/bsp/srf06eb_cc2538/drivers/source/bsp.h
index 38ac3d9..86efc35 100644
--- a/Components/bsp/srf06eb_cc2538/drivers/source/bsp.h
+++ b/Components/bsp/srf06eb_cc2538/drivers/source/bsp.h
@@ -81,38 +81,28 @@ extern "C"
 #define BSP_3V3_EN              GPIO_PIN_4      //!< PB4
 
 // Board LED defines
-#define BSP_LED_BASE            GPIO_C_BASE
-#define BSP_LED_1               GPIO_PIN_0      //!< PC0
+#define BSP_LED_BASE            GPIO_B_BASE
+#define BSP_LED_1               GPIO_PIN_1
 #define BSP_LED_2               GPIO_PIN_1      //!< PC1
 #define BSP_LED_3               GPIO_PIN_2      //!< PC2
 #define BSP_LED_4               GPIO_PIN_3      //!< PC3
-#define BSP_LED_ALL             (BSP_LED_1 | \
-                                 BSP_LED_2 | \
-                                 BSP_LED_3 | \
-                                 BSP_LED_4)     //!< Bitmask of all LEDs
+#define BSP_LED_ALL             BSP_LED_1
 
 // Board key defines
-#define BSP_KEY_DIR_BASE        GPIO_C_BASE     //!< Base for left/right/up/down
+#define BSP_KEY_DIR_BASE        GPIO_A_BASE
 #define BSP_KEY_SEL_BASE        GPIO_A_BASE     //!< Base for Select
 #define BSP_KEY_1               GPIO_PIN_4      //!< PC4
-#define BSP_KEY_2               GPIO_PIN_5      //!< PC5
+#define BSP_KEY_2							 GPIO_PIN_7
 #define BSP_KEY_3               GPIO_PIN_6      //!< PC6
 #define BSP_KEY_4               GPIO_PIN_7      //!< PC7
 #define BSP_KEY_5               GPIO_PIN_3      //!< PA3
-#define BSP_KEY_ALL             (BSP_KEY_1| \
-                                 BSP_KEY_2| \
-                                 BSP_KEY_3| \
-                                 BSP_KEY_4| \
-                                 BSP_KEY_5)     //!< Bitmask of all keys
+#define BSP_KEY_ALL             BSP_KEY_2
 #define BSP_KEY_LEFT            BSP_KEY_1
-#define BSP_KEY_RIGHT           BSP_KEY_2
+#define BSP_KEY_RIGHT         	BSP_KEY_2
 #define BSP_KEY_UP              BSP_KEY_3
 #define BSP_KEY_DOWN            BSP_KEY_4
 #define BSP_KEY_SELECT          BSP_KEY_5
-#define BSP_KEY_DIR_ALL         (BSP_KEY_LEFT|  \
-                                 BSP_KEY_RIGHT| \
-                                 BSP_KEY_UP|    \
-                                 BSP_KEY_DOWN)  //!< Bitmask of all dir. keys
+#define BSP_KEY_DIR_ALL         BSP_KEY_RIGHT
 
 // Board LCD defines
 #define BSP_LCD_MODE_BASE       GPIO_B_BASE
diff --git a/Components/hal/include/hal_led.h b/Components/hal/include/hal_led.h
index bb327f1..2892056 100644
--- a/Components/hal/include/hal_led.h
+++ b/Components/hal/include/hal_led.h
@@ -63,7 +63,7 @@ extern "C"
 #define HAL_LED_2     0x02
 #define HAL_LED_3     0x04
 #define HAL_LED_4     0x08
-#define HAL_LED_ALL   (HAL_LED_1 | HAL_LED_2 | HAL_LED_3 | HAL_LED_4)
+#define HAL_LED_ALL   (HAL_LED_1)
 
 /* Modes */
 #define HAL_LED_MODE_OFF     0x00
diff --git a/Components/hal/target/CC2538/hal_board_cfg.h b/Components/hal/target/CC2538/hal_board_cfg.h
index 4e4ecf0..9a45113 100644
--- a/Components/hal/target/CC2538/hal_board_cfg.h
+++ b/Components/hal/target/CC2538/hal_board_cfg.h
@@ -81,7 +81,7 @@
 
 #define xHAL_PA_LNA
 #define xHAL_PA_LNA_CC2590
-#define xHAL_PA_LNA_CC2592
+#define HAL_PA_LNA_CC2592
 
 /* HGM Port and Pin */
 #define HGM_BASE   GPIO_D_BASE
@@ -149,7 +149,7 @@
 #if defined (HAL_PA_LNA) || defined (HAL_PA_LNA_CC2592) || \
     defined (HAL_PA_LNA_CC2590) || defined (MAC_RUNTIME_CC2591) || \
     defined (MAC_RUNTIME_CC2592) || defined (MAC_RUNTIME_CC2590)
-#define HAL_NUM_LEDS            2
+#define HAL_NUM_LEDS            2       //1
 #else
 #define HAL_NUM_LEDS            4
 #endif
@@ -221,7 +221,7 @@
 #define HAL_FLASH_WORD_SIZE       4
 
 // Z-Stack uses flash pages for NV
-#define HAL_NV_PAGE_CNT           6
+#define HAL_NV_PAGE_CNT           12    //6
 #define HAL_NV_PAGE_BEG           (HAL_NV_PAGE_END-HAL_NV_PAGE_CNT+1)
 #define HAL_NV_START_ADDR         ((FLASH_BASE) + (HAL_NV_PAGE_BEG * HAL_FLASH_PAGE_SIZE))
 
@@ -362,26 +362,27 @@ extern void MAC_RfFrontendSetup(void);
                                                                             \
   /* Configure the leds as outputs */                                       \
   GPIOPinTypeGPIOOutput(BSP_LED_BASE, BSP_LED_1);                           \
-  GPIOPinTypeGPIOOutput(BSP_LED_BASE, BSP_LED_2);                           \
-  GPIOPinTypeGPIOOutput(BSP_LED_BASE, BSP_LED_3);                           \
-  GPIOPinTypeGPIOOutput(BSP_LED_BASE, BSP_LED_4);                           \
+  /*GPIOPinTypeGPIOOutput(BSP_LED_BASE, BSP_LED_2);*/                       \
+  /*GPIOPinTypeGPIOOutput(BSP_LED_BASE, BSP_LED_3); */                      \
+  /*GPIOPinTypeGPIOOutput(BSP_LED_BASE, BSP_LED_4); */                      \
                                                                             \
   /* reset does not affect GPIO state */                                    \
   HAL_TURN_OFF_LED1();                                                      \
-  HAL_TURN_OFF_LED2();                                                      \
-  HAL_TURN_OFF_LED3();                                                      \
-  HAL_TURN_OFF_LED4();                                                      \
+  /*HAL_TURN_OFF_LED2();*/                                                  \
+  /*HAL_TURN_OFF_LED3();*/                                                  \
+  /*HAL_TURN_OFF_LED4();*/                                                  \
                                                                             \
-  GPIOPinTypeGPIOOutput(BSP_LCD_PWR_BASE, BSP_LCD_PWR);  /* PWR output */   \
-  GPIOPinWrite(BSP_LCD_PWR_BASE, BSP_LCD_PWR, 0);        /* low */          \
-  GPIOPinTypeGPIOInput(BSP_LCD_CS_BASE, BSP_LCD_CS); /*CSn input no drive*/ \
+  /*GPIOPinTypeGPIOOutput(BSP_LCD_PWR_BASE, BSP_LCD_PWR);*/  /* PWR output */   \
+  /*GPIOPinWrite(BSP_LCD_PWR_BASE, BSP_LCD_PWR, 0);*/        /* low */          \
+  /*GPIOPinTypeGPIOInput(BSP_LCD_CS_BASE, BSP_LCD_CS);*/ /*CSn input no drive*/ \
                                                                             \
   INIT_PWM_PINS()                                                           \
                                                                             \
 }
 
 /* ----------- LED's ---------- */
-#define HAL_TURN_OFF_LED1()       GPIOPinWrite(BSP_LED_BASE, BSP_LED_1, 0)
+//#define HAL_TURN_OFF_LED1()       GPIOPinWrite(BSP_LED_BASE, BSP_LED_1, 0)
+#define HAL_TURN_OFF_LED1()       GPIOPinWrite(BSP_LED_BASE, BSP_LED_1, BSP_LED_1)
 #define HAL_TURN_OFF_LED2()       GPIOPinWrite(BSP_LED_BASE, BSP_LED_2, 0)
 
 #if defined (HAL_PA_LNA) || defined (HAL_PA_LNA_CC2592) || \
@@ -394,7 +395,8 @@ extern void MAC_RfFrontendSetup(void);
 #define HAL_TURN_OFF_LED4()       GPIOPinWrite(BSP_LED_BASE, BSP_LED_4, 0)
 #endif
 
-#define HAL_TURN_ON_LED1()        GPIOPinWrite(BSP_LED_BASE, BSP_LED_1, BSP_LED_1)
+//#define HAL_TURN_ON_LED1()        GPIOPinWrite(BSP_LED_BASE, BSP_LED_1, BSP_LED_1)
+#define HAL_TURN_ON_LED1()        GPIOPinWrite(BSP_LED_BASE, BSP_LED_1, 0)
 #define HAL_TURN_ON_LED2()        GPIOPinWrite(BSP_LED_BASE, BSP_LED_2, BSP_LED_2)
 #if defined (HAL_PA_LNA) || defined (HAL_PA_LNA_CC2592) || \
     defined (HAL_PA_LNA_CC2590) || defined (MAC_RUNTIME_CC2591) || \
@@ -458,7 +460,7 @@ extern void MAC_RfFrontendSetup(void);
 
 /* Set to TRUE enable LCD usage, FALSE disable it */
 #ifndef HAL_LCD
-#define HAL_LCD TRUE
+#define HAL_LCD FALSE   //TRUE
 #endif
 
 /* Set to TRUE enable LED usage, FALSE disable it */
diff --git a/Components/hal/target/CC2538/hal_key.c b/Components/hal/target/CC2538/hal_key.c
index b8359d7..bc71a2c 100644
--- a/Components/hal/target/CC2538/hal_key.c
+++ b/Components/hal/target/CC2538/hal_key.c
@@ -101,24 +101,24 @@ uint8 bspKeyPushedEx(uint8_t ui8ReadMask)
   uint32_t ui32DirPins;
   uint32_t ui32SelPin;
   uint32_t ui32DirPins_prev = ~GPIOPinRead(BSP_KEY_DIR_BASE, BSP_KEY_DIR_ALL);
-  uint32_t ui32SelPin_prev = ~GPIOPinRead(BSP_KEY_SEL_BASE, BSP_KEY_SELECT);
+  //uint32_t ui32SelPin_prev = ~GPIOPinRead(BSP_KEY_SEL_BASE, BSP_KEY_SELECT);
   uint16 i;
   
   // Read pin states + debounce
   for(i = 0; i < 500; i++)
   {
     ui32DirPins = ~GPIOPinRead(BSP_KEY_DIR_BASE, BSP_KEY_DIR_ALL);
-    ui32SelPin  = ~GPIOPinRead(BSP_KEY_SEL_BASE, BSP_KEY_SELECT);
+    //ui32SelPin  = ~GPIOPinRead(BSP_KEY_SEL_BASE, BSP_KEY_SELECT);
     
-    if ((ui32DirPins != ui32DirPins_prev) || (ui32SelPin != ui32SelPin_prev))
+    if ((ui32DirPins != ui32DirPins_prev)/* || (ui32SelPin != ui32SelPin_prev)*/)
     {
       i = 0;
       ui32DirPins_prev = ui32DirPins;
-      ui32SelPin_prev = ui32SelPin;
+      //ui32SelPin_prev = ui32SelPin;
     }
   }
   
-  return ((ui32DirPins & (BSP_KEY_LEFT | BSP_KEY_RIGHT | BSP_KEY_UP | BSP_KEY_DOWN) | (ui32SelPin & BSP_KEY_SELECT)));
+  return ((ui32DirPins & BSP_KEY_RIGHT/* (BSP_KEY_LEFT | BSP_KEY_RIGHT | BSP_KEY_UP | BSP_KEY_DOWN) | (ui32SelPin & BSP_KEY_SELECT)*/));
 }
 #endif
 
@@ -162,7 +162,7 @@ uint8 hal_key_keys(void)
 #endif
   if(ucKeysPressed & BSP_KEY_LEFT)
   {
-    x |= HAL_KEY_SW_4;
+    //x |= HAL_KEY_SW_4;
   }
   if(ucKeysPressed & BSP_KEY_RIGHT)
   {
@@ -170,15 +170,15 @@ uint8 hal_key_keys(void)
   }
   if(ucKeysPressed & BSP_KEY_UP)
   {
-    x |= HAL_KEY_SW_1;
+    //x |= HAL_KEY_SW_1;
   }
   if(ucKeysPressed & BSP_KEY_DOWN)
   {
-    x |= HAL_KEY_SW_3;
+    //x |= HAL_KEY_SW_3;
   }
   if(ucKeysPressed & BSP_KEY_SELECT)
   {
-    x |= HAL_KEY_SW_5;
+    //x |= HAL_KEY_SW_5;
   }
   return x; 
 }
@@ -206,7 +206,7 @@ uint8 hal_key_int_keys(void)
 
   if(ucKeysPressed & BSP_KEY_LEFT)
   {
-    x |= HAL_KEY_SW_4;
+    //x |= HAL_KEY_SW_4;
   }
   if(ucKeysPressed & BSP_KEY_RIGHT)
   {
@@ -214,11 +214,11 @@ uint8 hal_key_int_keys(void)
   }
   if(ucKeysPressed & BSP_KEY_UP)
   {
-    x |= HAL_KEY_SW_1;
+    //x |= HAL_KEY_SW_1;
   }
   if(ucKeysPressed & BSP_KEY_DOWN)
   {
-    x |= HAL_KEY_SW_3;
+    //x |= HAL_KEY_SW_3;
   }
   
   
@@ -230,7 +230,7 @@ uint8 hal_key_int_keys(void)
 
   if(ucKeysPressed & BSP_KEY_SELECT)
   {
-    x |= HAL_KEY_SW_5;
+    //x |= HAL_KEY_SW_5;
   }
   return x;
 }
@@ -262,16 +262,16 @@ void HalKeyConfig( bool interruptEnable, halKeyCBack_t cback)
     bspKeyInit(BSP_KEY_MODE_ISR);
     
     /* Map function dirKeyIsr to UP, LEFT, RIGHT and DOWN keys */
-    bspKeyIntRegister((BSP_KEY_UP|BSP_KEY_LEFT|BSP_KEY_RIGHT|BSP_KEY_DOWN),
+    bspKeyIntRegister(BSP_KEY_RIGHT,
                       &interrupt_keybd);
     
     /* Map function selectKeyIsr to SELECT key */
-    bspKeyIntRegister(BSP_KEY_SELECT, &interrupt_keybd);
+    //bspKeyIntRegister(BSP_KEY_SELECT, &interrupt_keybd);
     
     /* Enable interrupts on all keys */
     bspKeyIntEnable(BSP_KEY_ALL);
     
-    IntPrioritySet(INT_GPIOC, HAL_INT_PRIOR_KEY);              
+    //IntPrioritySet(INT_GPIOC, HAL_INT_PRIOR_KEY);              
     IntPrioritySet(INT_GPIOA, HAL_INT_PRIOR_KEY);     
     
     /* Cancel polling if there is one active */
@@ -394,12 +394,12 @@ uint8 hal_key_no_debounce(void)
   dirPins = (~GPIOPinRead(BSP_KEY_DIR_BASE,
                           BSP_KEY_DIR_ALL)) \
                           & BSP_KEY_DIR_ALL;
-  selPin  = (~GPIOPinRead(BSP_KEY_SEL_BASE,
+  /*selPin  = (~GPIOPinRead(BSP_KEY_SEL_BASE,
                           BSP_KEY_SELECT)) \
-                          & BSP_KEY_SELECT;
+                          & BSP_KEY_SELECT;*/
   if(dirPins & BSP_KEY_LEFT)
   {
-    x |= HAL_KEY_SW_4;
+    //x |= HAL_KEY_SW_4;
   }
   if(dirPins & BSP_KEY_RIGHT)
   {
@@ -407,15 +407,15 @@ uint8 hal_key_no_debounce(void)
   }
   if(dirPins & BSP_KEY_UP)
   {
-    x |= HAL_KEY_SW_1;
+    //x |= HAL_KEY_SW_1;
   }
   if(dirPins & BSP_KEY_DOWN)
   {
-    x |= HAL_KEY_SW_3;
+    //x |= HAL_KEY_SW_3;
   }
   if(selPin & BSP_KEY_SELECT)
   {
-    x |= HAL_KEY_SW_5;
+    //x |= HAL_KEY_SW_5;
   }
   return x;
 }
diff --git a/Components/hal/target/CC2538/startup_ewarm.c b/Components/hal/target/CC2538/startup_ewarm.c
index 83ae1b4..2b43d7e 100644
--- a/Components/hal/target/CC2538/startup_ewarm.c
+++ b/Components/hal/target/CC2538/startup_ewarm.c
@@ -55,6 +55,7 @@
 #else 
 #define FLASH_START_ADDR 0x00200000
 #define BOOTLOADER_BACKDOOR_DISABLE     0xEFFFFFFF
+#define BOOTLOADER_BACKDOOR_ENABLE			0xF7FFFFFF
 #endif
 
 //
@@ -140,7 +141,7 @@ typedef struct
 
 __root const lockPageCCA_t __cca @ ".cca" =
 {
-  BOOTLOADER_BACKDOOR_DISABLE,  // Bootloader backdoor disabled
+  BOOTLOADER_BACKDOOR_ENABLE,  // Bootloader backdoor 
   0,                            // Image valid bytes
   FLASH_START_ADDR // Vector table located at flash start address
 };
diff --git a/Components/osal/mcu/cc2538/osal_nv.c b/Components/osal/mcu/cc2538/osal_nv.c
index 9275e64..6d18224 100644
--- a/Components/osal/mcu/cc2538/osal_nv.c
+++ b/Components/osal/mcu/cc2538/osal_nv.c
@@ -62,7 +62,7 @@
  * $PROJ_DIR$\..\..\..\Tools\"Processor Specific Name"\"Specific Name".xcl
  */
 #ifndef OSAL_NV_PHY_PER_PG
-  #define OSAL_NV_PHY_PER_PG    1
+  #define OSAL_NV_PHY_PER_PG    2//1
 #endif
 
 #define OSAL_NV_PAGES_USED     (HAL_NV_PAGE_CNT / OSAL_NV_PHY_PER_PG)
diff --git a/Components/stack/nwk/nwk_globals.c b/Components/stack/nwk/nwk_globals.c
index eb436ae..1741289 100644
--- a/Components/stack/nwk/nwk_globals.c
+++ b/Components/stack/nwk/nwk_globals.c
@@ -71,10 +71,10 @@
  * CONSTANTS
  */
 // Maximums for the data buffer queue
-#define NWK_MAX_DATABUFS_WAITING    8     // Waiting to be sent to MAC
-#define NWK_MAX_DATABUFS_SCHEDULED  5     // Timed messages to be sent
-#define NWK_MAX_DATABUFS_CONFIRMED  5     // Held after MAC confirms
-#define NWK_MAX_DATABUFS_TOTAL      12    // Total number of buffers
+#define NWK_MAX_DATABUFS_WAITING    180//8     // Waiting to be sent to MAC
+#define NWK_MAX_DATABUFS_SCHEDULED  150//5     // Timed messages to be sent
+#define NWK_MAX_DATABUFS_CONFIRMED  150//5     // Held after MAC confirms
+#define NWK_MAX_DATABUFS_TOTAL      255//12    // Total number of buffers
 
 // 1-255 (0 -> 256) X RTG_TIMER_INTERVAL
 // A known shortcoming is that when a message is enqueued as "hold" for a
diff --git a/Components/stack/sys/ZGlobals.c b/Components/stack/sys/ZGlobals.c
index d170283..7ff7645 100644
--- a/Components/stack/sys/ZGlobals.c
+++ b/Components/stack/sys/ZGlobals.c
@@ -212,7 +212,7 @@ uint8 zgEndDeviceConfiguration = END_DEV_CONFIGURATION;
 //
 // NOTICE:  Before enabling Child Aging make sure to review all the related
 // definitions in this file, especially zgNwkParentInformation.
-uint8 zgChildAgingEnable = TRUE;
+uint8 zgChildAgingEnable = FALSE;	//TRUE;
 
 //==========    TouchLink NWK configuration    ===============
 // Values used by Router when starts a network as initiator
diff --git a/Components/zmac/ZMAC.h b/Components/zmac/ZMAC.h
index d658113..356ecc6 100644
--- a/Components/zmac/ZMAC.h
+++ b/Components/zmac/ZMAC.h
@@ -457,8 +457,11 @@ typedef enum
   TX_PWR_PLUS_15,
   TX_PWR_PLUS_16,
   TX_PWR_PLUS_17,
-  TX_PWR_PLUS_18,
-  TX_PWR_PLUS_19
+  TX_PWR_PLUS_18, 
+  TX_PWR_PLUS_19,
+  TX_PWR_PLUS_20,
+  TX_PWR_PLUS_21,
+  TX_PWR_PLUS_22
 } ZMacTransmitPower_t;  // The transmit power in units of -1 dBm.
 
 typedef struct
diff --git a/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp.c b/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp.c
index a6b89c9..69f9307 100644
--- a/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp.c
+++ b/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp.c
@@ -118,7 +118,7 @@ byte zclGenericApp_TaskID;
 /*********************************************************************
  * LOCAL VARIABLES
  */
-
+void Report_ID_Attr(void);
 uint8 giGenAppScreenMode = GENERIC_MAINMODE;   // display the main screen mode first
 
 uint8 gPermitDuration = 0;    // permit joining default to disabled
@@ -243,6 +243,8 @@ void zclGenericApp_Init( byte task_id )
 {
   zclGenericApp_TaskID = task_id;
 
+  ZMacSetTransmitPower(TX_PWR_PLUS_22);
+  
   // This app is part of the Home Automation Profile
   bdb_RegisterSimpleDescriptor( &zclGenericApp_SimpleDesc );
 
@@ -291,8 +293,10 @@ void zclGenericApp_Init( byte task_id )
 #ifdef LCD_SUPPORTED
   HalLcdWriteString ( (char *)sDeviceName, HAL_LCD_LINE_3 );
 #endif  // LCD_SUPPORTED
-
-
+  bdb_StartCommissioning(BDB_COMMISSIONING_MODE_NWK_STEERING |
+                         BDB_COMMISSIONING_MODE_FINDING_BINDING);
+  
+  osal_start_timerEx(zclGenericApp_TaskID, GENERICAPP_EVT_2, 500);
 }
 
 /*********************************************************************
@@ -333,8 +337,15 @@ uint16 zclGenericApp_event_loop( uint8 task_id, uint16 events )
                (zclGenericApp_NwkState == DEV_ROUTER)   ||
                (zclGenericApp_NwkState == DEV_END_DEVICE) )
           {
-            giGenAppScreenMode = GENERIC_MAINMODE;
-            zclGenericApp_LcdDisplayUpdate();
+            HalLedSet ( HAL_LED_1, HAL_LED_MODE_ON );
+            osal_stop_timerEx(zclGenericApp_TaskID, GENERICAPP_EVT_2);
+            Report_ID_Attr();
+          }
+          if(zclGenericApp_NwkState == DEV_HOLD) {
+            bdb_StartCommissioning(BDB_COMMISSIONING_MODE_NWK_FORMATION | 
+                             BDB_COMMISSIONING_MODE_NWK_STEERING | 
+                             BDB_COMMISSIONING_MODE_FINDING_BINDING | 
+                             BDB_COMMISSIONING_MODE_INITIATOR_TL);
           }
           break;
 
@@ -370,21 +381,29 @@ uint16 zclGenericApp_event_loop( uint8 task_id, uint16 events )
   
   
   if ( events & GENERICAPP_EVT_1 )
-  {
-    // toggle LED 2 state, start another timer for 500ms
-    HalLedSet ( HAL_LED_2, HAL_LED_MODE_TOGGLE );
-    osal_start_timerEx( zclGenericApp_TaskID, GENERICAPP_EVT_1, 500 );
-    
+  {   
+    if (bdbAttributes.bdbNodeIsOnANetwork) {
+      bdb_resetLocalAction();
+    } else {
+      bdb_StartCommissioning(BDB_COMMISSIONING_MODE_NWK_FORMATION | 
+                             BDB_COMMISSIONING_MODE_NWK_STEERING | 
+                             BDB_COMMISSIONING_MODE_FINDING_BINDING | 
+                             BDB_COMMISSIONING_MODE_INITIATOR_TL);
+      
+      osal_start_timerEx(zclGenericApp_TaskID, GENERICAPP_EVT_2, 500);
+    }
     return ( events ^ GENERICAPP_EVT_1 );
   }
   
-  /*
+  
   if ( events & GENERICAPP_EVT_2 )
   {
-    
+    // toggle LED 2 state, start another timer for 500ms
+    HalLedSet ( HAL_LED_1, HAL_LED_MODE_TOGGLE );
+    osal_start_timerEx( zclGenericApp_TaskID, GENERICAPP_EVT_2, 500 );
     return ( events ^ GENERICAPP_EVT_2 );
   }
-  
+  /*
   if ( events & GENERICAPP_EVT_3 )
   {
     
@@ -395,7 +414,30 @@ uint16 zclGenericApp_event_loop( uint8 task_id, uint16 events )
   // Discard unknown events
   return 0;
 }
+afAddrType_t strRepId;
+uint16 SeqNum = 0;
+void Report_ID_Attr(void) {
+  zclReportCmd_t *pReportCmd;
 
+  pReportCmd = osal_mem_alloc(sizeof(zclReportCmd_t) +
+                              (sizeof(zclReport_t)));
+  if (pReportCmd != NULL) {
+    pReportCmd->numAttr = 1;
+    pReportCmd->attrList[0].attrID = ATTRID_BASIC_MODEL_ID;
+    pReportCmd->attrList[0].dataType = ZCL_DATATYPE_CHAR_STR;
+    pReportCmd->attrList[0].attrData = (void *)(&zclGenericApp_ModelId);
+
+    strRepId.addrMode = (afAddrMode_t)Addr16Bit;
+    strRepId.addr.shortAddr = 0;
+    strRepId.endPoint = 1;
+
+    zcl_SendReportCmd(GENERICAPP_ENDPOINT, &strRepId,
+                      ZCL_CLUSTER_ID_GEN_BASIC, pReportCmd,
+                      ZCL_FRAME_CLIENT_SERVER_DIR, false, SeqNum++);
+  }
+
+  osal_mem_free(pReportCmd);
+}
 
 /*********************************************************************
  * @fn      zclGenericApp_HandleKeys
@@ -412,65 +454,18 @@ uint16 zclGenericApp_event_loop( uint8 task_id, uint16 events )
  * @return  none
  */
 static void zclGenericApp_HandleKeys( byte shift, byte keys )
-{
-  if ( keys & HAL_KEY_SW_1 )
-  {
-    static bool LED_OnOff = FALSE;
-    
-    giGenAppScreenMode = GENERIC_MAINMODE;
-    
-    /* GENERICAPP_TODO: add app functionality to hardware keys here */
-    
-    // for example, start/stop LED 2 toggling with 500ms period
-    if (LED_OnOff)
-    { 
-      // if the LED is blinking, stop the osal timer and turn the LED off
-      osal_stop_timerEx(zclGenericApp_TaskID, GENERICAPP_EVT_1);
-      HalLedSet ( HAL_LED_2, HAL_LED_MODE_OFF );
-      LED_OnOff = FALSE;
-    }
-    else
-    {
-      // turn on LED 2 and start an osal timer to toggle it after 500ms, search
-      // for GENERICAPP_EVT_1 to see event handling after expired timer
-      osal_start_timerEx( zclGenericApp_TaskID, GENERICAPP_EVT_1, 500 );
-      HalLedSet ( HAL_LED_2, HAL_LED_MODE_ON );
-      LED_OnOff = TRUE;
-    }
-  }
+{ 
   // Start the BDB commissioning method
   if ( keys & HAL_KEY_SW_2 )
-  {
-    giGenAppScreenMode = GENERIC_MAINMODE;
-
-    bdb_StartCommissioning(BDB_COMMISSIONING_MODE_NWK_FORMATION | BDB_COMMISSIONING_MODE_NWK_STEERING | BDB_COMMISSIONING_MODE_FINDING_BINDING | BDB_COMMISSIONING_MODE_INITIATOR_TL);
-  }
-  if ( keys & HAL_KEY_SW_3 )
-  {
-    giGenAppScreenMode = GENERIC_MAINMODE;
-  
-    // touchlink target commissioning, if enabled  
-#if ( defined ( BDB_TL_TARGET ) && (BDB_TOUCHLINK_CAPABILITY_ENABLED == TRUE) )
-    bdb_StartCommissioning(BDB_COMMISSIONING_MODE_FINDING_BINDING);
-    touchLinkTarget_EnableCommissioning( 30000 );
-#endif
-    
+  {   
+    osal_start_timerEx(zclGenericApp_TaskID, GENERICAPP_EVT_1, 5000 );
+    HalLedSet ( HAL_LED_1, HAL_LED_MODE_BLINK );
+    Report_ID_Attr();
+    return;
+    //bdb_StartCommissioning(BDB_COMMISSIONING_MODE_NWK_FORMATION | BDB_COMMISSIONING_MODE_NWK_STEERING | BDB_COMMISSIONING_MODE_FINDING_BINDING | BDB_COMMISSIONING_MODE_INITIATOR_TL);
+  } else {
+    osal_stop_timerEx(zclGenericApp_TaskID, GENERICAPP_EVT_1);    
   }
-  if ( keys & HAL_KEY_SW_4 )
-  {
-    
-   giGenAppScreenMode = giGenAppScreenMode ? GENERIC_MAINMODE : GENERIC_HELPMODE;
-#ifdef LCD_SUPPORTED
-    HalLcdWriteString( (char *)sClearLine, HAL_LCD_LINE_2 );
-#endif
-    
-  }
-  if ( keys & HAL_KEY_SW_5 )
-  {
-    bdb_resetLocalAction();
-  }
-
-  zclGenericApp_LcdDisplayUpdate();
 }
 
 /*********************************************************************
@@ -640,11 +635,11 @@ static void zclGenericApp_ProcessIdentifyTimeChange( uint8 endpoint )
 
   if ( zclGenericApp_IdentifyTime > 0 )
   {
-    HalLedBlink ( HAL_LED_2, 0xFF, HAL_LED_DEFAULT_DUTY_CYCLE, HAL_LED_DEFAULT_FLASH_TIME );
+    HalLedBlink ( HAL_LED_1, 0xFF, HAL_LED_DEFAULT_DUTY_CYCLE, HAL_LED_DEFAULT_FLASH_TIME );
   }
   else
   {
-    HalLedSet ( HAL_LED_2, HAL_LED_MODE_OFF );
+    HalLedSet ( HAL_LED_1, HAL_LED_MODE_ON );
   }
 }
 
diff --git a/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp.h b/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp.h
index 47c3bdc..df6a2a8 100644
--- a/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp.h
+++ b/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp.h
@@ -61,7 +61,7 @@ extern "C"
 /*********************************************************************
  * CONSTANTS
  */
-#define GENERICAPP_ENDPOINT            8
+#define GENERICAPP_ENDPOINT            1
 // Added to include ZLL Target functionality
 #define GENERICAPP_NUM_GRPS            2
 
@@ -74,8 +74,9 @@ extern "C"
 /* GENERICAPP_TODO: define app events here */
   
 #define GENERICAPP_EVT_1                    0x0008
-/*
+
 #define GENERICAPP_EVT_2                    0x0010
+ /*
 #define GENERICAPP_EVT_3                    0x0020
 */
 
@@ -110,6 +111,7 @@ extern CONST uint8 zclCmdsArraySize;
 // attribute list
 extern CONST zclAttrRec_t zclGenericApp_Attrs[];
 extern CONST uint8 zclGenericApp_NumAttributes;
+extern CONST uint8 zclGenericApp_ModelId[];
 
 // Identify attributes
 extern uint16 zclGenericApp_IdentifyTime;
diff --git a/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp_data.c b/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp_data.c
index d0d6ee2..2719700 100644
--- a/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp_data.c
+++ b/Projects/zstack/HomeAutomation/GenericApp/Source/zcl_genericapp_data.c
@@ -93,9 +93,9 @@ const uint16 zclGenericApp_clusterRevision_all = 0x0001;
 // Basic Cluster
 const uint8 zclGenericApp_HWRevision = GENERICAPP_HWVERSION;
 const uint8 zclGenericApp_ZCLVersion = GENERICAPP_ZCLVERSION;
-const uint8 zclGenericApp_ManufacturerName[] = { 16, 'T','e','x','a','s','I','n','s','t','r','u','m','e','n','t','s' };
-const uint8 zclGenericApp_ModelId[] = { 16, 'T','I','0','0','0','1',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ' };
-const uint8 zclGenericApp_DateCode[] = { 16, '2','0','0','6','0','8','3','1',' ',' ',' ',' ',' ',' ',' ',' ' };
+const uint8 zclGenericApp_ManufacturerName[] = { 7, 'j','e','t','h','o','m','e' };
+const uint8 zclGenericApp_ModelId[] = { 16, 'c','c','2','5','3','8','.','r','o','u','t','e','r','.','v','1' };
+const uint8 zclGenericApp_DateCode[] = { 16, '2','0','2','0','0','5','2','0',' ',' ',' ',' ',' ',' ',' ',' ' };
 const uint8 zclGenericApp_PowerSource = POWER_SOURCE_MAINS_1_PHASE;
 
 uint8 zclGenericApp_LocationDescription[17] = { 16, ' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ' };
@@ -162,7 +162,7 @@ CONST zclAttrRec_t zclGenericApp_Attrs[] =
     { // Attribute record
       ATTRID_BASIC_MODEL_ID,
       ZCL_DATATYPE_CHAR_STR,
-      ACCESS_CONTROL_READ,
+      (ACCESS_CONTROL_READ | ACCESS_REPORTABLE),
       (void *)zclGenericApp_ModelId
     }
   },
diff --git a/Projects/zstack/Tools/CC2538DB/CC2538.icf b/Projects/zstack/Tools/CC2538DB/CC2538.icf
index d4c95cc..e077a3f 100644
--- a/Projects/zstack/Tools/CC2538DB/CC2538.icf
+++ b/Projects/zstack/Tools/CC2538DB/CC2538.icf
@@ -15,7 +15,7 @@ define memory mem with size = 4G;
 //
 // Define a region for the on-chip program code space.
 //
-define region FLASH = mem:[from 0x00200000 to 0x0027C7FF];
+define region FLASH = mem:[from 0x00200000 to 0x002797FF];	//mem:[from 0x00200000 to 0x0027C7FF];
 
 //
 // Define a region for the OTA CRC structure.
@@ -39,7 +39,7 @@ define region FLASH = mem:[from 0x00200000 to 0x0027C7FF];
 //   The size of this region MUST MATCH the size defined by "HAL_NV_PAGE_CNT"
 //   in the file: hal_board_cfg.h
 //
-define region NV_MEM = mem:[from 0x0027C800 to 0x0027F7FF];
+define region NV_MEM = mem:[from 0x00279800 to 0x0027F7FF];		//mem:[from 0x0027C800 to 0x0027F7FF];
 
 //
 // Define regions for on-chip factory Commissioning Parameters.
@@ -87,7 +87,7 @@ define region FLASH_LCK = mem:[from 0x0027FFE0 to 0x0027FFFF];
 //
 // Define a region for the on-chip SRAM.
 //
-define region SRAM = mem:[from 0x20004000 to 0x20007FFF];
+define region SRAM = mem:[from 0x20000000 to 0x20007FFF];		//mem:[from 0x20004000 to 0x20007FFF];
 
 //
 // Define a block for the heap.  The size should be set to something other
diff --git a/Projects/zstack/Tools/CC2538DB/f8wConfig.cfg b/Projects/zstack/Tools/CC2538DB/f8wConfig.cfg
index 9d2c1bf..3bfbcaa 100644
--- a/Projects/zstack/Tools/CC2538DB/f8wConfig.cfg
+++ b/Projects/zstack/Tools/CC2538DB/f8wConfig.cfg
@@ -48,7 +48,7 @@
 //-DDEFAULT_CHANLIST=0x00004000  // 14 - 0x0E
 //-DDEFAULT_CHANLIST=0x00002000  // 13 - 0x0D
 //-DDEFAULT_CHANLIST=0x00001000  // 12 - 0x0C
--DDEFAULT_CHANLIST=0x00000800  // 11 - 0x0B
+-DDEFAULT_CHANLIST=0x07FFF800    //0x00000800  // 11 - 0x0B
 
 /* Define the default PAN ID.
  *
diff --git a/Projects/zstack/Tools/CC2538DB/f8wRouter.cfg b/Projects/zstack/Tools/CC2538DB/f8wRouter.cfg
index 713a1ef..b468a95 100644
--- a/Projects/zstack/Tools/CC2538DB/f8wRouter.cfg
+++ b/Projects/zstack/Tools/CC2538DB/f8wRouter.cfg
@@ -16,5 +16,41 @@
 -DMAC_CFG_TX_MAX=8
 -DMAC_CFG_RX_MAX=5
 
+-DINT_HEAP_LEN=12288
+-DNWK_MAX_DEVICE_LIST=100
+-DMAX_RTG_SRC_ENTRIES=400
+-DMAX_NEIGHBOR_ENTRIES=100
+-DMTO_RREQ_LIMIT_TIME=5000
+
+-DLINK_DOWN_TRIGGER=6
+-DNWK_ROUTE_AGE_LIMIT=12
+-DBCAST_DELIVERY_TIME=100
+-DMAX_BCAST=12
+-DROUTE_DISCOVERY_TIME=13
+-DNWK_LINK_STATUS_PERIOD=60
 /* Router Settings */
 -DRTR_NWK                          // Router Functions
+
+-DCC2538_USE_ALTERNATE_INTERRUPT_MAP=1
+-DFEATURE_RESET_MACRO
+-DSECURE=1
+-DTC_LINKKEY_JOIN
+-DNV_INIT
+-DNV_RESTORE
+//-DZTOOL_P1
+//-DMT_TASK
+//-DMT_APP_FUNC
+//-DMT_SYS_FUNC
+//-DMT_ZDO_FUNC
+//-DMT_APP_CNF_FUNC
+//-DLEGACY_LCD_DEBUG
+//-DLCD_SUPPORTED=DEBUG
+-DMULTICAST_ENABLED=FALSE
+-DZCL_READ
+-DZCL_WRITE
+-DZCL_BASIC
+-DZCL_IDENTIFY
+-DZCL_SCENES
+-DZCL_GROUPS
+-DDISABLE_GREENPOWER_BASIC_PROXY
+-DZCL_REPORTING_DEVICE
\ No newline at end of file
